src/libcore/ops/drop.rs

```diff
@@ -1,26 +1,26 @@
-/// Used to run some code when a value goes out of scope.
-/// This is sometimes called a 'destructor'.
+/// Destructor for cleaning up resources.
 ///
-/// When a value goes out of scope, it will have its `drop` method called if
-/// its type implements `Drop`. Then, any fields the value contains will also
-/// be dropped recursively.
+/// When a value is no longer needed, Rust will run a "destructor" on that value.
+/// The most common way that a value is no longer needed is when it goes out of
+/// scope. Destructors may still run in other circumstances, but we're going to
+/// focus on scope for the examples here. To learn about some of those other cases,
+/// please see [the reference] section on destructors.
 ///
-/// Because of this recursive dropping, you do not need to implement this trait
-/// unless your type needs its own destructor logic.
+/// [the reference]: https://doc.rust-lang.org/reference/destructors.html
 ///
-/// Refer to [the chapter on `Drop` in *The Rust Programming Language*][book]
-/// for some more elaboration.
-///
-/// [book]: ../../book/ch15-03-drop.html
-///
-/// # Examples
+/// The `Drop` trait provides a way to implement a custom destructor on a type.
+/// While this is not necessary for most types, there are some cases where
+/// it is useful. Most notably when dealing with resources. That resource may
+/// be memory, it may be a file descriptor, it may be a network socket.
+/// But when the type is no longer going to be used, it should "clean up" that
+/// resource by freeing the memory or closing the file or socket. This is
+/// the job of a destructor, and therefore the job of `Drop::drop`.
 ///
 /// ## Implementing `Drop`
 ///
-/// The `drop` method is called when `_x` goes out of scope, and therefore
-/// `main` prints `Dropping!`.
+/// When a value goes out of scope, the compiler automatically insert a call to `Drop::drop` on that value. For example,
```

Comment:
Maybe refer to something like `drop_in_place` instead of `Drop::drop` directly here. By implementing `Drop::drop` the behavior of `drop_in_place` is customized. cc. @RalfJung

Reply:
hm, I don't want to also use `drop_in_place`, even if it is slightly more correct. :thinking: 

For new users the difference seems to be negligible and causes them to look at one more place. We already have `Drop::drop` and `mem::drop` here, which is more complicated than most people expect when first learning about `drop`.

Reply:
I think the text here is incorrect though, and misleading. If anything it should say that a call to `mem::drop` is inserted.

For example, when you have your own type like
```rust
struct Foo(Vec<u8>);
```
and you `impl Drop for Foo`, then calling `Drop::drop` would *not* drop the vector! `Drop::drop` is shallow, `mem::drop`/`drop_in_place` are deep. I do think this is something that must be explained.

IMO it makes more sense to talk about `Drop::drop` and `drop_in_place`, and leave out `mem::drop` (it can be explained as an afterthought, as a simple safe wrapper around `drop_in_place`). `Drop::drop` drops "one layer" of data structures, and `drop_in_place` adds a bunch of auto-generated code to recursively drop each layer of nested structs/enums.

Reply:
The more I think about this the less I wan to use `drop_in_place` here. This method is unsafe and its documentation is not really beginner friendly (which is ok, as no beginner should have to use it tbh).

> and you impl Drop for Foo, then calling Drop::drop would not drop the vector! Drop::drop is shallow, mem::drop/drop_in_place are deep. I do think this is something that must be explained.

This is explained in the section `Dropping is done recursively` further down. I would like to move this section upwards, but the section `You cannot call Drop::drop yourself` is also important, so I don't want to move that one further down. :sweat_smile:

Reply:
> This method is unsafe and its documentation is not really beginner friendly (which is ok, as no beginner should have to use it tbh).

Maybe its doc should be improved? IMO properly understanding drop requires understanding `drop_in_place`. That might not be beginner-friendly, but that is because dropping is actually non-trivial.

Saying "the compiler automatically insert a call to `Drop::drop`" is factually incorrect and thus I think we have to say something else. I appreciate sometimes it's helpful to say something "almost correct" to ease understanding (I do that all the time), but what it currently says is too wrong to be useful, IMO. Going into `drop_in_place` is just one way to fix that (it's what I'd do, but I am open to alternatives).

Reply:
My current ideas which use  `mem::drop` instead of `Drop::drop` are either more incorrect or more complex. :thinking:

Reply:
IMO saying "it automatically inserts calls to `mem::drop`" would be less incorrect than what it says now. `mem::drop` is "move followed by `drop_in_place`"; this would "just" be "lying" about the move part -- in contrast to the current text which entirely conflates user-written drop impls with auto-generated drop glue.

Reply:
> I appreciate sometimes it's helpful to say something "almost correct" to ease understanding (I do that all the time when writing papers), but what it currently says is too wrong to be useful, IMO.

hmm, how about

---


When a value goes out of scope, the compiler automatically insert a call to `Drop::drop` for that value, and then recursively for all other values it contains:

```rust
struct HasDrop;

impl Drop for HasDrop {
    fn drop(&mut self) {
        println!("Dropping HasDrop!");
    }
}

struct HasTwoDrops {
    one: HasDrop,
    two: HasDrop,
}

impl Drop for HasTwoDrops {
    fn drop(&mut self) {
        println!("Dropping HasTwoDrops!");
    }
}

fn main() {
    let _x = HasTwoDrops { one: HasDrop, two: HasDrop };
}
```

Running this will print

```text
Dropping HasTwoDrops!
Dropping HasDrop!
Dropping HasDrop!
```

Even if we remove the implementation of `Drop` for `HasTwoDrop`, the destructors of its fields
are still called. This would result in

```test
Dropping HasDrop!
Dropping HasDrop!
```

Reply:
> When a value goes out of scope, the compiler automatically insert a call to Drop::drop for that value, and then recursively for all other values it contains

That is at least not incorrect any more. :D 

But I still think the docs should *name* that concept of "call Drop::drop for that value, and then recursively for all other values it contains". Naming things makes explaining and remembering them much easier. The name I have used  recentlyfor that concept is `drop_in_place`. Sometimes it's also called "drop glue", though I am not sure if that is strictly the same thing (maybe "drop glue" is just the code that does the gluing, as opposed to the entire operation). I am open for other suggestions for what this should be called, but I feel rather strongly that it should be called *something*.

Reply:
I now use "drop glue" there, updated the PR. 

> But I still think the docs should name that concept of "call Drop::drop for that value, and then recursively for all other values it contains".

I think that I personally use "dropping" for this, i.e. when "dropping" a value, the compiler calls `Drop::drop` for it and all its owned values.

Reply:
> I now use "drop glue" there, updated the PR.

I like that terminology. :D
I'll leave some specific comments about the structure.